ed25519.go

```diff
@@ -0,0 +1,170 @@
+	}
+	pub = make([]byte, publicKeySizeEd25519)
+	copy(pub, priv[seedSizeEd25519:])
+	return pub, priv, nil
+}
+
+func NewKeyFromSeedEd25519(seed []byte) (priv []byte, err error) {
+	// Outline the function body so that the returned key can be stack-allocated.
+	priv = make([]byte, privateKeySizeEd25519)
+	err = newKeyFromSeedEd25519(priv, seed)
+	if err != nil {
+		return nil, err
+	}
+	return priv, err
+}
+
+func newKeyFromSeedEd25519(priv []byte, seed []byte) error {
+	if l := len(seed); l != seedSizeEd25519 {
+		panic("ed25519: bad seed length: " + strconv.Itoa(l))
+	}
+	pkey := C.go_openssl_EVP_PKEY_new_raw_private_key(C.GO_EVP_PKEY_ED25519, nil, base(seed), C.size_t(len(seed)))
+	if pkey == nil {
+		return newOpenSSLError("EVP_PKEY_new_raw_private_key")
+	}
+	defer C.go_openssl_EVP_PKEY_free(pkey)
+	return extractPKEYPrivEd25519(pkey, priv)
+}
+
+func extractPKEYPrivEd25519(pkey C.GO_EVP_PKEY_PTR, priv []byte) error {
+	pubSize, privSize := C.size_t(publicKeySizeEd25519), C.size_t(seedSizeEd25519)
+	if C.go_openssl_EVP_PKEY_get_raw_public_key(pkey, base(priv[seedSizeEd25519:]), &pubSize) != 1 {
+		return newOpenSSLError("EVP_PKEY_get_raw_public_key")
+	}
+	if pubSize != publicKeySizeEd25519 {
+		panic("ed25519: bad public key length: " + strconv.Itoa(int(pubSize)))
+	}
+	if C.go_openssl_EVP_PKEY_get_raw_private_key(pkey, base(priv), &privSize) != 1 {
+		return newOpenSSLError("EVP_PKEY_get_raw_private_key")
+	}
+	if privSize != seedSizeEd25519 {
+		panic("ed25519: bad private key length: " + strconv.Itoa(int(privSize)))
+	}
+	return nil
+}
+
+func SignEd25519(priv, message []byte) (sig []byte, err error) {
+	// Outline the function body so that the returned key can be stack-allocated.
+	sig = make([]byte, signatureSizeEd25519)
+	err = signEd25519(sig, priv, message)
+	if err != nil {
+		return nil, err
+	}
+	return sig, err
+}
+
+func signEd25519(sig, priv, message []byte) error {
+	if l := len(priv); l != privateKeySizeEd25519 {
+		panic("ed25519: bad private key length: " + strconv.Itoa(l))
+	}
+	pkey := C.go_openssl_EVP_PKEY_new_raw_private_key(C.GO_EVP_PKEY_ED25519, nil, base(priv[:seedSizeEd25519]), seedSizeEd25519)
+	if pkey == nil {
+		return newOpenSSLError("EVP_PKEY_new_raw_private_key")
+	}
+	defer C.go_openssl_EVP_PKEY_free(pkey)
+	ctx := C.go_openssl_EVP_MD_CTX_new()
+	if ctx == nil {
+		return newOpenSSLError("EVP_MD_CTX_new")
+	}
+	defer C.go_openssl_EVP_MD_CTX_free(ctx)
+	if C.go_openssl_EVP_DigestSignInit(ctx, nil, nil, nil, pkey) != 1 {
+		return newOpenSSLError("EVP_DigestSignInit")
+	}
+	siglen := C.size_t(signatureSizeEd25519)
+	if C.go_openssl_EVP_DigestSign(ctx, base(sig), &siglen, base(message), C.size_t(len(message))) != 1 {
+		return newOpenSSLError("EVP_DigestSign")
+	}
+	if siglen != signatureSizeEd25519 {
+		panic("ed25519: bad signature length: " + strconv.Itoa(int(siglen)))
+	}
+	return nil
+}
+
+func VerifyEd25519(pub, message, sig []byte) error {
+	return verifyEd25519(pub, message, sig)
+}
```

Comment:
Why is this one a wrapper even though nothing's being outlined?

Reply:
It's a left-over, I'll fix that.